packages/contracts-bedrock/src/Safe/LivenessGuard.sol

```diff
@@ -0,0 +1,158 @@
+
+    /// @notice A mapping of the timestamp at which an owner last participated in signing a
+    ///         an executed transaction, or called showLiveness.
+    mapping(address => uint256) public lastLive;
+
+    /// @notice An enumerable set of addresses used to store the list of owners before execution,
+    ///         and then to update the lastLive mapping according to changes in the set observed
+    ///         after execution.
+    EnumerableSet.AddressSet private ownersBefore;
+
+    /// @notice Constructor.
+    /// @param _safe The safe account for which this contract will be the guard.
+    constructor(Safe _safe) {
+        SAFE = _safe;
+        address[] memory owners = _safe.getOwners();
+        for (uint256 i = 0; i < owners.length; i++) {
+            address owner = owners[i];
+            lastLive[owner] = block.timestamp;
+            emit OwnerRecorded(0x0, owner);
+        }
+    }
+
+    /// @notice Internal function to ensure that only the Safe can call certain functions.
+    function _onlySafe() internal view {
+        require(msg.sender == address(SAFE), "LivenessGuard: only Safe can call this function");
+    }
+
+    /// @notice Records the most recent time which any owner has signed a transaction.
+    /// @dev Called by the Safe contract before execution of a transaction.
+    function checkTransaction(
+        address to,
+        uint256 value,
+        bytes memory data,
+        Enum.Operation operation,
+        uint256 safeTxGas,
+        uint256 baseGas,
+        uint256 gasPrice,
+        address gasToken,
+        address payable refundReceiver,
+        bytes memory signatures,
+        address msgSender
+    )
+        external
+    {
+        msgSender; // silence unused variable warning
+        _onlySafe();
+
+        // Cache the set of owners prior to execution.
+        // This will be used in the checkAfterExecution method.
+        address[] memory owners = SAFE.getOwners();
+        uint256 ownersLength = owners.length;
+        for (uint256 i = 0; i < ownersLength; i++) {
+            ownersBefore.add(owners[i]);
+        }
+
+        // This call will reenter to the Safe which is calling it. This is OK because it is only reading the
+        // nonce, and using the getTransactionHash() method.
+        bytes32 txHash = Safe(payable(msg.sender)).getTransactionHash({
+            to: to,
+            value: value,
+            data: data,
+            operation: operation,
+            safeTxGas: safeTxGas,
+            baseGas: baseGas,
+            gasPrice: gasPrice,
+            gasToken: gasToken,
+            refundReceiver: refundReceiver,
+            _nonce: Safe(payable(msg.sender)).nonce() - 1
+        });
+
+        uint256 threshold = SAFE.getThreshold();
+        address[] memory signers =
+            SafeSigners.getNSigners({ dataHash: txHash, signatures: signatures, requiredSignatures: threshold });
+
+        for (uint256 i = 0; i < signers.length; i++) {
```

Comment:
We cache array length in the above loop but not here, but we should be consistent. Personally I'd inline the above length instead of caching to remove code and make it more readable, since the gas savings will be negligible, but this contradicts @tynes suggestion and I'm ok with either

Reply:
Ok I am fine with either, I am in favor of caching in all cases but we should favor consistency

Reply:
I've elected to not cache.